### ISSUE TITLE
Can't construct rvalue reference to array type during dereference

### ISSUE THREAD

**Issue by fbrausse** [esbmc/esbmc#1943]
This error still occurs, but we don't have access to the source (industrial partner).
Backtrace:
```
segfault_handler(int) at esbmc_parseoptions.cpp:?
__restore_rt at libc_sigaction.o:?
pthread_kill at ??:?
gsignal at ??:?
abort at ??:?
dereferencet::build_reference_rec(irep_container<expr2t>&, irep_container<expr2t> const&, irep_container<type2t> const&, guardt const&, dereferencet::modet, unsigned long) [clone .localalias] at dereference.cpp:?
dereferencet::build_reference_to(irep_container<expr2t> const&, dereferencet::modet, irep_container<expr2t> const&, irep_container<type2t> const&, guardt const&, irep_container<expr2t> const&, irep_container<expr2t>&) at ??:?
dereferencet::dereference(irep_container<expr2t> const&, irep_container<type2t> const&, guardt const&, dereferencet::modet, irep_container<expr2t> const&) at ??:?
dereferencet::dereference_expr_nonscalar(irep_container<expr2t>&, guardt&, dereferencet::modet, irep_container<expr2t> const&) at ??:?
dereferencet::dereference_expr_nonscalar(irep_container<expr2t>&, guardt&, dereferencet::modet, irep_container<expr2t> const&) at ??:?
dereferencet::dereference_expr(irep_container<expr2t>&, guardt&, dereferencet::modet) at ??:?
void call_expr_delegate<std::vector<irep_container<expr2t>, std::allocator<irep_container<expr2t> > >, std::function<void (irep_container<expr2t>&)> >(std::vector<irep_container<expr2t>, std::allocator<irep_container<expr2t> > >&, std::function<void (irep_container<expr2t>&)>&) at ??:?
dereferencet::dereference_expr(irep_container<expr2t>&, guardt&, dereferencet::modet) at ??:?
goto_symext::dereference(irep_container<expr2t>&, dereferencet::modet) at ??:?
goto_symext::symex_assign(irep_container<expr2t> const&, bool, guardt const&) at ??:?
execution_statet::symex_assign(irep_container<expr2t> const&, bool, guardt const&) at ??:?
goto_symext::symex_step(reachability_treet&) at ??:?
reachability_treet::get_next_formula() at ??:?
bmct::run_thread(std::shared_ptr<symex_target_equationt>&) at ??:?
bmct::run(std::shared_ptr<symex_target_equationt>&) at ??:?
bmct::start_bmc() at ??:?
esbmc_parseoptionst::do_bmc(bmct&) at ??:?
esbmc_parseoptionst::is_base_case_violated(optionst&, goto_functionst&, BigInt const&) at ??:?
esbmc_parseoptionst::do_bmc_strategy(optionst&, goto_functionst&) at ??:?
esbmc_parseoptionst::doit() at ??:?
main at ??:?
__libc_start_call_main at libc-start.o:?
__libc_start_main_impl at ??:?
_start at ??:?
```
For value:
```
symbol
* name : __ESBMC_main::$tmp::tmp$77
* renamelev : Level 1
* level1_num : 1
* level2_num : 0
* thread_num : 0
* node_num : 0
* type : struct
  * members :
    0: array
      * subtype : unsignedbv
          * width : 8
      * array_size : constant_int
          * value : 3
          * type : unsignedbv
            * width : 64
      * size_is_infinite : false
    1: unsignedbv
      * width : 8
    2: array
      * subtype : unsignedbv
          * width : 32
      * array_size : constant_int
          * value : 3
          * type : unsignedbv
            * width : 64
      * size_is_infinite : false
    3: array
      * subtype : unsignedbv
          * width : 32
      * array_size : constant_int
          * value : 3
          * type : unsignedbv
            * width : 64
      * size_is_infinite : false
    4: array
      * subtype : unsignedbv
          * width : 8
      * array_size : constant_int
          * value : 3
          * type : unsignedbv
            * width : 64
      * size_is_infinite : false
    5: bool
    6: unsignedbv
      * width : 8
    7: unsignedbv
      * width : 8
    8: unsignedbv
      * width : 8
    9: unsignedbv
      * width : 8
    10: unsignedbv
      * width : 16
    11: unsignedbv
      * width : 16
    12: unsignedbv
      * width : 32
    13: floatbv
      * fraction : 23
      * exponent : 8
    14: unsignedbv
      * width : 8
    15: unsignedbv
      * width : 8
    16: unsignedbv
      * width : 8
    17: unsignedbv
      * width : 8
    18: unsignedbv
      * width : 16
    19: unsignedbv
      * width : 16
    20: unsignedbv
      * width : 32
    21: floatbv
      * fraction : 23
      * exponent : 8
    22: unsignedbv
      * width : 32
 
  * member_names :
    0: is_any_leaf_rapl_limited
    1: anon_pad$1
    2: current_rapl_perf_status
    3: prev_tsc_rapl_perf_status
    4: domain_vr_id
    5: sst_cp_input_available
    6: psys_pbm_limit
    7: core_pbm_limit
    8: clos_pbm_limit
    9: anon_pad$9
    10: dram_pbm_limit
    11: dram_domain_freq_limit
    12: max_limit_scaling_factor
    13: pid_min_limit
    14: min_pid_output_domain
    15: min_pid_output_instance
    16: min_pid_output_interface
    17: anon_pad$17
    18: limiting_reason_mask
    19: anon_pad$19
    20: PERF_LIMIT_REASON_NUM_OF_BITS
    21: core_pbm_limit_scaled
    22: total_num_cores
 
  * member_pretty_names :
    0: is_any_leaf_rapl_limited
    1: anon_pad$1
    2: current_rapl_perf_status
    3: prev_tsc_rapl_perf_status
    4: domain_vr_id
    5: sst_cp_input_available
    6: psys_pbm_limit
    7: core_pbm_limit
    8: clos_pbm_limit
    9: anon_pad$9
    10: dram_pbm_limit
    11: dram_domain_freq_limit
    12: max_limit_scaling_factor
    13: pid_min_limit
    14: min_pid_output_domain
    15: min_pid_output_instance
    16: min_pid_output_interface
    17: anon_pad$17
    18: limiting_reason_mask
    19: anon_pad$19
    20: PERF_LIMIT_REASON_NUM_OF_BITS
    21: core_pbm_limit_scaled
    22: total_num_cores
 
  * typename : Primecode::RAPLVars
  * packed : false
```
type:
```
array
* subtype : unsignedbv
    * width : 8
* array_size : constant_int
    * value : 3
    * type : unsignedbv
      * width : 64
* size_is_infinite : false
```
offset:
```
constant_int
* value : 0
* type : unsignedbv
  * width : 67
```
mode: read aligned
alignment: 8

----------------------------------

It's interesting that this appears to happen for code introduces in `__ESBMC_main`, according to the symbol's name `__ESBMC_main::$tmp::tmp$77`. Any ideas where `$tmp::tmp$77` might be created?

**Comment by fbrausse** [comment]
This might just be a case of [C11 6.7.2.1p15](http://port70.net/%7Ensz/c/c11/n1570.html#6.7.2.1p15):
> A pointer to a structure object, suitably converted, points to its initial member (or if that member is a bit-field, then to the unit in which it resides), and vice versa. There may be unnamed padding within a structure object, but not at its beginning.

I'm pretty sure this also applies to C++ POD types, for compatibility with C...

If so, I think the dereference module should allow this particular case.

**Comment by fbrausse** [comment]
Fixed by #1968 as well, thanks @XLiZHI and @ChenfengWei0.
